src/Components/Selection.re

```diff
@@ -0,0 +1,75 @@
+[@deriving show({with_path: false})]
+type t = {
+  anchor: int,
+  focus: int,
+};
+
+type directionVariant =
+  | Left(int)
+  | Right(int);
+
+let initial: t = {anchor: 0, focus: 0};
+
+let withinLength = (text: string, position: int): int => {
+  let withingStart = max(position, 0);
+  let withingBoth = min(withingStart, String.length(text));
+
+  withingBoth;
+};
+
+let create = (~text: string, ~anchor: int, ~focus: int): t => {
+  let safeAnchor = withinLength(text, anchor);
+  let safeFocus = withinLength(text, focus);
+
+  {anchor: safeAnchor, focus: safeFocus};
+};
+
+let anchor = (selection: t): int => {
+  selection.anchor;
+};
+
+let focus = (selection: t): int => {
+  selection.focus;
+};
+
+let range = (selection: t): int => {
+  abs(selection.focus - selection.anchor);
+};
+
+let rangeStart = (selection: t): int => {
+  min(selection.focus, selection.anchor);
+};
+
+let rangeEnd = (selection: t): int => {
+  max(selection.focus, selection.anchor);
+};
+
+let isCollapsed = (selection: t): bool => {
+  selection.anchor == selection.focus;
+};
+
+let collapse = (~text: string, offset: int): t => {
+  create(~text, ~anchor=offset, ~focus=offset);
+};
```

Comment:
The name of this indicates that it will collapse a selection, when in fact it creates a new selection that is collapsed. Perhaps just rename it to `collapsed`?

Reply:
I was inspired this thing https://developer.mozilla.org/en-US/docs/Web/API/Selection
For browser it does make since it mutate object.
In this case I agree. `collapsed` make much more sence.